### PR TITLE
New harder task

### PR DESCRIPTION
**Pull request by josephdviviano**
- [x] I've read the [.github/CONTRIBUTING.md](.github/CONTRIBUTING.md) file
- [x] My code follows the typing guidelines
- [x] I've added appropriate tests
- [x] I've run pre-commit hooks locally

## Description

+ Added 3 new hypergrid tasks which should be more challenging. Note that the specifics are very much up for debate. I tried to identify environments which were easy to divide + conquer vs those which require compositional knowledge (and therefore some amount of knowledge sharing among agents in a multi-agent setting).
+ Added mode verification logic (to ensure that your particular configuration actually contains modes to find).
+ Added lots of tests around these new rewards.
+ Added visualizations of the reward landscape for these various rewards.

### CODE REVIEW

File: src/gfn/gym/hypergrid.py [hunk 1]
```diff
@@ -375,6 +537,342 @@ def n_terminating_states(self) -> int:
...
+    def _exists_fallback_random(self, thr: float) -> bool:
+        """Random sampling fallback.
+
+        Draw a modest batch of random states on CPU and accept if any exceed the
+        threshold with a small tolerance. This is a last resort to avoid
+        expensive enumeration for large grids.
+        """
+        with torch.no_grad():
+            device = torch.device("cpu")
+            B = min(2048, max(128, 8 * self.ndim))
```
Comment (younik) [inline src/gfn/gym/hypergrid.py:831]: what are these numbers? Maybe use constant to improve clarity

File: src/gfn/gym/hypergrid.py [hunk 2]
```diff
@@ -375,6 +537,342 @@ def n_terminating_states(self) -> int:
...
+            if self.n_states <= 200_000:
+                axes = [
+                    torch.arange(self.height, dtype=torch.long) for _ in range(self.ndim)
+                ]
+                grid = torch.cartesian_prod(*axes)
+                rewards = self.reward_fn(grid)
+                return bool((rewards >= thr - EPS_REWARD_CMP).any().item())
+        except Exception:
+            # Fall back to heuristic paths below
+            pass
```
[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/gym/hypergrid.py:574]: yes, agreed

**Comment by josephdviviano** [comment]
Hi @younik - I hear you, this is a big PR. The "splits" would have to be along tasks, though, so the resulting PRs would still be large.

I appreciate your comments on the code. I think it would make sense to also look at the tasks (the stuff that's plotted in the notebook) to see if they make sense. I'm not convinced by all of the tasks.

I would be open to removing a task or two. I think the one that works best for it's intended purpose is the coprime reward.

File: src/gfn/gym/hypergrid.py [hunk 3]
```diff
@@ -320,9 +438,53 @@ def modes_found(self, states: DiscreteStates) -> set[int]:
...
+        if self._mode_stats_kind == "exact" and self._n_modes_via_ids_exact is not None:
+            return int(self._n_modes_via_ids_exact)
+        if (
+            self._mode_stats_kind == "approx"
+            and self._n_modes_via_ids_estimate is not None
+        ):
+            return int(self._n_modes_via_ids_estimate)
+
         return 2**self.ndim
```
[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/gym/hypergrid.py:488]: no you're right it should be stored.

**Comment by hyeok9855** [comment]
> I would be open to removing a task or two. I think the one that works best for it's intended purpose is the coprime reward.

I do think `Template Minkowski` and `Bitwise/XOR` rewards are not very interesting to benchmark, especially if you care about the mode coverage. `Multiplicative/Coprime` seems challenging, but you may want to increase the reward for further modes from the origin.

**Review by saleml — CHANGES_REQUESTED** [review]
This is high-quality research code with excellent mathematical foundations and thorough testing. The main concerns are:

- Complexity barrier for new users
- Performance documentation gaps
- Some missing edge-case handling

A few questions and suggestions

- The new reward functions are mathematically sophisticated (GF(2) algebra, prime factorization, etc.). While excellent for research, the barrier to entry is high
Suggestion: Add a "Quick Start" section to the documentation showing simple use cases before diving into the mathematical details.
- The `_solve_gf2_has_solution` method uses Gaussian elimination which could be slow for large constraint systems
Suggestion: Add performance warnings in docstrings
- Why GF(2)?  The choice is elegant but not obvious. Could you add a paragraph in the documentation explaining why linear algebra over GF(2) is natural for compositional structure?
-  What happens if a user picks "impossible" preset with ndim=2, height=16? Should the factory functions validate compatibility?
- Can you consider adding type hings for kwargs? something like
```
class BitwiseXORRewardKwargs(TypedDict, total=False):
    R0: float
    tier_weights: list[float]
    dims_constrained: list[int]
    bits_per_tier: list[tuple[int, int]]
    parity_checks: list[dict] | None
```
- What do you think of adding visualization helpers
```
def visualize_mode_structure(env: HyperGrid, sample_size: int = 10000):
    """Generate 2D/3D plots of mode distribution."""
    # Auto-generate plots similar to notebook but as API
```


I'd also like to suggest a structural consideration:
The original HyperGrid has become a pedagogical entrypoint of the GFlowNets library:

- It's the first environment new users encounter
- Its simplicity (grid + distance-based reward) makes it ideal for teaching core concepts
- Tutorial code often uses it as the "Hello World" of GFlowNets
- The cognitive load is intentionally minimal: "navigate a grid, reach high-reward corners"

Can we consider creating a separate file src/gfn/gym/compositional_hypergrid.py that:
- Inherits from HyperGrid to reuse the core grid mechanics
- Houses the new reward families (BitwiseXOR, MultiplicativeCoprime, TemplateMinkowski)
- Includes the sophisticated mode validation and statistics machinery
- Keeps the original simple and focused on accessibility
